--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -3112,6 +3112,7 @@ static int airoha_probe(struct platform_
 	printk("\n [AIROHA_ETH]Successfully Identify CHIP ID:%d \n",eth->airoha_chip_id);
 
 	eth->meter_enable = of_property_read_bool(eth->dev->of_node, "meter-enable");
+	eth->multicast_meteridx = of_property_read_bool(dev->of_node,"multicast-meteridx");
 	
 	err = dma_set_mask_and_coherent(eth->dev, DMA_BIT_MASK(32));
 	if (err) {
--- a/drivers/net/ethernet/airoha/airoha_eth.h
+++ b/drivers/net/ethernet/airoha/airoha_eth.h
@@ -755,6 +755,7 @@ struct airoha_eth {
 	struct qdma_init qdma_init;
 	bool work_started;
 	bool meter_enable;
+	unsigned int multicast_meteridx;
 };
 
 u32 airoha_rr(void __iomem *base, u32 offset);
