--- a/drivers/net/ethernet/airoha/airoha_ppe.c
+++ b/drivers/net/ethernet/airoha/airoha_ppe.c
@@ -10,13 +10,23 @@
 #include <linux/platform_device.h>
 #include <linux/rhashtable.h>
 #include <net/ipv6.h>
+#include <linux/if_pppox.h>
 #include <net/pkt_cls.h>
 
 #include "airoha_regs.h"
 #include "airoha_eth.h"
 
+extern int airoha_set_ppe_mac(struct airoha_foe_entry *foe_entry, struct net_device *dev, char* src_mac, char*dst_mac, u16 pppid);
+extern int (*ra_sw_nat_hook_clean_table) (void);
+extern void airoha_ppe_hook_init(void);
+extern void airoha_ppe_hook_exit(void);
+
+extern struct airoha_eth *glb_eth;
+extern int airoha_is_pon_sfu_mode(void);
+
 static DEFINE_MUTEX(flow_offload_mutex);
-static DEFINE_SPINLOCK(ppe_lock);
+DEFINE_SPINLOCK(ppe_lock);
+static DEFINE_SPINLOCK(ppe_sram_access_lock);
 
 static const struct rhashtable_params airoha_flow_table_params = {
 	.head_offset = offsetof(struct airoha_flow_table_entry, node),
@@ -96,6 +106,9 @@ static void airoha_ppe_hw_init(struct ai
 	sram_tb_size = sram_num_entries * sizeof(struct airoha_foe_entry);
 	dram_num_entries = PPE_RAM_NUM_ENTRIES_SHIFT(PPE_DRAM_NUM_ENTRIES);
 
+	airoha_fe_rmw(eth, PSE_FC_CFG, PSE_PORT7_SHARE_DIS,
+			FIELD_PREP(PSE_PORT7_SHARE_DIS, 0));
+
 	sram_num_stats_entries = airoha_ppe_get_num_stats_entries(ppe);
 	if (sram_num_stats_entries > 0)
 		sram_ppe_num_data_entries -= sram_num_stats_entries;
@@ -139,12 +152,19 @@ static void airoha_ppe_hw_init(struct ai
 			      PPE_TB_CFG_KEEPALIVE_MASK |
 			      PPE_TB_ENTRY_SIZE_MASK,
 			      FIELD_PREP(PPE_TB_CFG_SEARCH_MISS_MASK, 3) |
+			      FIELD_PREP(PPE_TB_CFG_KEEPALIVE_MASK, 0) |
 			      FIELD_PREP(PPE_TB_ENTRY_SIZE_MASK, 0) |
 			      FIELD_PREP(PPE_SRAM_TB_NUM_ENTRY_MASK,
 					 sram_ppe_num_data_entries) |
 			      FIELD_PREP(PPE_DRAM_TB_NUM_ENTRY_MASK,
 					 dram_num_entries));
 
+		airoha_fe_rmw(eth, REG_PPE_BIND_RATE(i),
+			      PPE_BIND_RATE_L2B_BIND_MASK |
+			      PPE_BIND_RATE_BIND_MASK,
+			      FIELD_PREP(PPE_BIND_RATE_L2B_BIND_MASK, 0x1E) |
+			      FIELD_PREP(PPE_BIND_RATE_BIND_MASK, 0x1E));
+
 		airoha_fe_wr(eth, REG_PPE_HASH_SEED(i), PPE_HASH_SEED);
 
 		for (p = 0; p < ARRAY_SIZE(eth->ports); p++)
@@ -155,6 +175,10 @@ static void airoha_ppe_hw_init(struct ai
 						 AIROHA_MAX_MTU) |
 				      FIELD_PREP(FP1_EGRESS_MTU_MASK,
 						 AIROHA_MAX_MTU));
+		airoha_fe_clear(eth, REG_PPE_PPE_FLOW_CFG(i),
+				PPE_FLOW_CFG_IP4_DSLITE_MASK |
+				PPE_FLOW_CFG_IP6_6RD_MASK);
+		
 	}
 }
 
@@ -266,6 +290,29 @@ static void airoha_ppe_foe_set_bridge_ad
 	br->src_mac_lo = get_unaligned_be32(eh->h_source + 2);
 }
 
+static inline int airoha_ppe_is_pon_dev(struct net_device *dev)
+{
+	return ((dev != NULL) && (dev->name[0] == 'p') && \
+		(dev->name[1] == 'o') && (dev->name[2] == 'n'));
+}
+
+static inline int airoha_ppe_is_ppp_dev(struct net_device *dev)
+{
+	return ((dev != NULL) && (dev->name[0] == 'p') && \
+		(dev->name[1] == 'p') && (dev->name[2] == 'p'));
+}
+
+static inline struct airoha_gdm_port *airoha_ppe_get_gdm_port
+	(struct airoha_eth *eth, struct net_device *dev)
+{
+	if(airoha_ppe_is_pon_dev(dev) ||
+		airoha_ppe_is_ppp_dev(dev)){
+		return eth->ports[1];
+	}
+
+	return (struct airoha_gdm_port *)netdev_priv(dev);
+}
+
 static int airoha_ppe_foe_entry_prepare(struct airoha_eth *eth,
 					struct airoha_foe_entry *hwe,
 					struct net_device *dev, int type,
@@ -290,21 +337,25 @@ static int airoha_ppe_foe_entry_prepare(
 
 	val = FIELD_PREP(AIROHA_FOE_IB2_PORT_AG, 0x1f);
 	if (dev) {
+		int id;
 		struct airoha_wdma_info info = {};
 
 		if (!airoha_ppe_get_wdma_info(dev, data->eth.h_dest, &info)) {
 			val |= FIELD_PREP(AIROHA_FOE_IB2_NBQ, info.idx) |
 			       FIELD_PREP(AIROHA_FOE_IB2_PSE_PORT,
 					  FE_PSE_PORT_CDM4);
+			val |= AIROHA_FOE_IB2_FAST_PATH;		  
 			qdata |= FIELD_PREP(AIROHA_FOE_ACTDP, info.bss);
 			wlan_etype = FIELD_PREP(AIROHA_FOE_MAC_WDMA_BAND,
 						info.idx) |
 				     FIELD_PREP(AIROHA_FOE_MAC_WDMA_WCID,
 						info.wcid);
 		} else {
-			struct airoha_gdm_port *port = netdev_priv(dev);
+			struct airoha_gdm_port *port;
 			u8 pse_port;
 
+			port = airoha_ppe_get_gdm_port(eth, dev);
+
 			if (!airoha_is_valid_gdm_port(eth, port))
 				return -EINVAL;
 
@@ -328,7 +379,33 @@ static int airoha_ppe_foe_entry_prepare(
 						  dsa_port);
 
 			smac_id = port->id;
-		}
+
+			id = port->qdma - &eth->qdma[0];
+			if (dsa_port >= 0) {
+				if (eth->qdma_init.lan_fastpath == 1) {
+					val |= AIROHA_FOE_IB2_FAST_PATH;
+				} else {
+					val &= ~AIROHA_FOE_IB2_FAST_PATH;
+				}
+			} else {
+				if (id == 0) {
+					if(eth->qdma_init.xsi_ether_fastpath == 1) {
+						val |= AIROHA_FOE_IB2_FAST_PATH;
+					} else {
+						val &= ~AIROHA_FOE_IB2_FAST_PATH;
+					}
+				} else if (id == 1) {
+					if(eth->qdma_init.wan_fastpath == 1) {
+						val |= AIROHA_FOE_IB2_FAST_PATH;
+					} else {
+						val &= ~AIROHA_FOE_IB2_FAST_PATH;
+					}
+				} else {
+					pr_err("%s Error\n", __func__);
+					return -EINVAL;
+				}	
+			}
+		}	
 	}
 
 	if (is_multicast_ether_addr(data->eth.h_dest))
@@ -357,22 +434,7 @@ static int airoha_ppe_foe_entry_prepare(
 		l2->etype = ETH_P_IP;
 	}
 
-	l2->dest_mac_hi = get_unaligned_be32(data->eth.h_dest);
-	l2->dest_mac_lo = get_unaligned_be16(data->eth.h_dest + 4);
-	if (type <= PPE_PKT_TYPE_IPV4_DSLITE) {
-		struct airoha_foe_mac_info *mac_info;
-
-		l2->src_mac_hi = get_unaligned_be32(data->eth.h_source);
-		hwe->ipv4.l2.src_mac_lo =
-			get_unaligned_be16(data->eth.h_source + 4);
-
-		mac_info = (struct airoha_foe_mac_info *)l2;
-		mac_info->pppoe_id = data->pppoe.sid;
-	} else {
-		l2->src_mac_hi = FIELD_PREP(AIROHA_FOE_MAC_SMAC_ID, smac_id) |
-				 FIELD_PREP(AIROHA_FOE_MAC_PPPOE_ID,
-					    data->pppoe.sid);
-	}
+	airoha_set_ppe_mac(hwe, dev, data->eth.h_source, data->eth.h_dest, data->pppoe.sid);
 
 	if (data->vlan.num) {
 		l2->vlan1 = data->vlan.hdr[0].id;
@@ -454,8 +516,8 @@ static int airoha_ppe_foe_entry_set_ipv6
 	return 0;
 }
 
-static u32 airoha_ppe_foe_get_entry_hash(struct airoha_ppe *ppe,
-					 struct airoha_foe_entry *hwe)
+u32 airoha_ppe_foe_get_entry_hash(struct airoha_ppe *ppe,
+				  struct airoha_foe_entry *hwe)
 {
 	int type = FIELD_GET(AIROHA_FOE_IB1_BIND_PACKET_TYPE, hwe->ib1);
 	u32 ppe_hash_mask = airoha_ppe_get_total_num_entries(ppe) - 1;
@@ -592,14 +654,14 @@ static void airoha_ppe_foe_flow_stats_up
 		      AIROHA_FOE_IB2_PSE_QOS | AIROHA_FOE_IB2_FAST_PATH);
 	*meter |= FIELD_PREP(AIROHA_FOE_TUNNEL_MTU, val);
 
-	nbq = pse_port == 1 ? 6 : 5;
+	nbq = hash < PPE_SRAM_NUM_ENTRIES ? 5 : 6;
 	*ib2 &= ~(AIROHA_FOE_IB2_NBQ | AIROHA_FOE_IB2_PSE_PORT |
 		  AIROHA_FOE_IB2_PSE_QOS);
 	*ib2 |= FIELD_PREP(AIROHA_FOE_IB2_PSE_PORT, 6) |
 		FIELD_PREP(AIROHA_FOE_IB2_NBQ, nbq);
 }
 
-static struct airoha_foe_entry *
+struct airoha_foe_entry *
 airoha_ppe_foe_get_entry_locked(struct airoha_ppe *ppe, u32 hash)
 {
 	u32 sram_num_entries = airoha_ppe_get_total_sram_num_entries(ppe);
@@ -643,7 +705,7 @@ struct airoha_foe_entry *airoha_ppe_foe_
 	return hwe;
 }
 
-static bool airoha_ppe_foe_compare_entry(struct airoha_flow_table_entry *e,
+bool airoha_ppe_foe_compare_entry(struct airoha_flow_table_entry *e,
 					 struct airoha_foe_entry *hwe)
 {
 	int type = FIELD_GET(AIROHA_FOE_IB1_BIND_PACKET_TYPE, e->data.ib1);
@@ -681,6 +743,60 @@ static int airoha_ppe_foe_commit_sram_en
 					REG_PPE_RAM_CTRL(ppe2));
 }
 
+
+struct airoha_flow_table_entry *airoha_flow_table_entry_get_by_hash(struct airoha_ppe *ppe, u32 hash){
+	struct airoha_flow_table_entry *e;
+	struct airoha_foe_entry *hwe = NULL;
+	struct hlist_node *n;
+	u32 index = 0;
+	if (hash > airoha_ppe_get_total_num_entries(ppe) - 1)
+		return NULL;
+	hwe = airoha_ppe_foe_get_entry_locked(ppe, hash);
+	if (!hwe)
+		return NULL;
+	index = airoha_ppe_foe_get_entry_hash(ppe, hwe);
+	hlist_for_each_entry_safe(e, n, &ppe->foe_flow[index], list) {
+		if(airoha_ppe_foe_compare_entry(e, hwe))
+			return e;
+	}
+	return NULL;
+}
+
+static int __airoha_ppe_foe_commit_entry(struct airoha_ppe *ppe, u32 hash)
+{
+	u32 sram_entries_num = airoha_ppe_get_total_sram_num_entries(ppe);
+	int ret = -1;
+	
+	lockdep_assert_held(&ppe_lock);
+	spin_lock_bh(&ppe_sram_access_lock);
+        if (likely(hash < sram_entries_num)) {
+                u32 *hwe = ppe->foe + hash * sizeof(struct airoha_foe_entry);
+                struct airoha_eth *eth = ppe->eth;
+                bool ppe2;
+                u32 val;
+                int i;
+
+                ppe2 = airoha_ppe_is_enabled(ppe->eth, 1) &&
+                       hash >= PPE_SRAM_NUM_ENTRIES;
+
+		for (i = 0; i < sizeof(struct airoha_foe_entry) / 4; i++)
+			airoha_fe_wr(eth, REG_PPE_RAM_ENTRY(ppe2, i), hwe[i]);
+
+		wmb();
+		airoha_fe_wr(ppe->eth, REG_PPE_RAM_CTRL(ppe2),
+                             FIELD_PREP(PPE_SRAM_CTRL_ENTRY_MASK, hash) |
+			     PPE_SRAM_CTRL_WR_MASK |
+                             PPE_SRAM_CTRL_REQ_MASK);
+		ret = read_poll_timeout_atomic(airoha_fe_rr, val,
+                                             val & PPE_SRAM_CTRL_ACK_MASK,
+                                             10, 100, false, eth,
+                                             REG_PPE_RAM_CTRL(ppe2));
+	}
+
+	spin_unlock_bh(&ppe_sram_access_lock);
+	return ret;
+}
+
 static int airoha_ppe_foe_commit_entry(struct airoha_ppe *ppe,
 				       struct airoha_foe_entry *e,
 				       u32 hash, bool rx_wlan)
@@ -697,6 +813,10 @@ static int airoha_ppe_foe_commit_entry(s
 
 	e->ib1 &= ~AIROHA_FOE_IB1_BIND_TIMESTAMP;
 	e->ib1 |= FIELD_PREP(AIROHA_FOE_IB1_BIND_TIMESTAMP, ts);
+	if(FIELD_GET(AIROHA_FOE_IB1_BIND_STATE, hwe->ib1) == AIROHA_FOE_STATE_UNBIND){
+		e->ib1 &= ~AIROHA_FOE_IB1_BIND_STATE;
+		e->ib1 |= FIELD_PREP(AIROHA_FOE_IB1_BIND_STATE, AIROHA_FOE_STATE_BIND);
+	}
 	hwe->ib1 = e->ib1;
 
 	rcu_read_lock();
@@ -711,12 +831,13 @@ static int airoha_ppe_foe_commit_entry(s
 		airoha_ppe_foe_flow_stats_update(ppe, npu, hwe, hash);
 
 	if (hash < sram_num_entries)
-		err = airoha_ppe_foe_commit_sram_entry(ppe, hash);
+		err = __airoha_ppe_foe_commit_entry(ppe, hash);
 unlock:
 	rcu_read_unlock();
 
 	return err;
 }
+int (*airoha_ppe_foe_commit_entry_ptr)(struct airoha_ppe *ppe,struct airoha_foe_entry *e,u32 hash,bool rx_wlan)=airoha_ppe_foe_commit_entry;
 
 static void airoha_ppe_foe_remove_flow(struct airoha_ppe *ppe,
 				       struct airoha_flow_table_entry *e)
@@ -806,8 +927,8 @@ airoha_ppe_foe_commit_subflow_entry(stru
 			memcpy(&hwe.ipv4.new_tuple, &hwe.ipv4.orig_tuple,
 			       sizeof(hwe.ipv4.new_tuple));
 	}
-
 	hwe.bridge.data = e->data.bridge.data;
+
 	airoha_ppe_foe_commit_entry(ppe, &hwe, hash, rx_wlan);
 
 	return 0;
@@ -848,9 +969,18 @@ static void airoha_ppe_foe_insert_entry(
 		if (!airoha_ppe_foe_compare_entry(e, hwe))
 			continue;
 
+		if(e->e_magic == PON && !(e->tx_modified)){
+			AIROHA_LOG(AIROHA_DEBUG_LEVEL_ERR, "PON upstream hash %u tx_modified is %u\n", hash, e->tx_modified);
+			goto unlock;
+		}else if((e->e_magic >= ETH2 && e->e_magic <= LAN4) && !(e->tx_modified)){
+			AIROHA_LOG(AIROHA_DEBUG_LEVEL_ERR, "Eth upstream hash %u tx_modified is %u\n", hash, e->tx_modified);
+			goto unlock;
+		}
+		//airoha_ppe_foe_commit_entry(ppe, &e->data, hash);
 		airoha_ppe_foe_commit_entry(ppe, &e->data, hash, rx_wlan);
 		commit_done = true;
 		e->hash = hash;
+		break;
 	}
 
 	if (commit_done)
@@ -999,6 +1129,27 @@ static int airoha_ppe_entry_idle_time(st
 	return airoha_ppe_get_entry_idle_time(ppe, e->data.ib1);
 }
 
+static void airoha_set_flow_destination(struct net_device *dev, struct airoha_flow_table_entry *e)
+{
+        if(!dev || !e){
+                return;
+        }
+        AIROHA_LOG(AIROHA_DEBUG_LEVEL_ERR, "device is %s\n",dev->name);
+
+        if(airoha_ppe_is_pon_dev(dev)){
+                AIROHA_LOG(AIROHA_DEBUG_LEVEL_ERR, "destination set to PON\n");
+                e->e_magic = PON;
+        }
+		
+		if(airoha_get_lan_id(dev))
+		{
+			AIROHA_LOG(AIROHA_DEBUG_LEVEL_ERR, "destination set to ETH\n");
+            e->e_magic = airoha_get_lan_id(dev);
+		}
+
+        return;
+}
+
 static int airoha_ppe_flow_offload_replace(struct airoha_eth *eth,
 					   struct flow_cls_offload *f)
 {
@@ -1100,8 +1251,9 @@ static int airoha_ppe_flow_offload_repla
 		}
 	}
 
-	if (!is_valid_ether_addr(data.eth.h_source) ||
-	    !is_valid_ether_addr(data.eth.h_dest))
+	if (!airoha_ppe_is_ppp_dev(odev) && 
+		(!is_valid_ether_addr(data.eth.h_source) ||
+	    !is_valid_ether_addr(data.eth.h_dest)))
 		return -EINVAL;
 
 	err = airoha_ppe_foe_entry_prepare(eth, &hwe, odev, offload_type,
@@ -1180,6 +1332,8 @@ static int airoha_ppe_flow_offload_repla
 	e->cookie = f->cookie;
 	memcpy(&e->data, &hwe, sizeof(e->data));
 
+	airoha_set_flow_destination(odev, e);
+
 	err = airoha_ppe_foe_flow_commit_entry(eth->ppe, e);
 	if (err)
 		goto free_entry;
@@ -1416,6 +1570,66 @@ void airoha_ppe_init_upd_mem(struct airo
 		     FIELD_PREP(PPE_UPDMEM_ADDR_MASK, port->id) |
 		     FIELD_PREP(PPE_UPDMEM_OFFSET_MASK, 1) |
 		     PPE_UPDMEM_WR_MASK | PPE_UPDMEM_REQ_MASK);
+	UpdateShrinkTable(port->id,addr);
+
+}
+
+void airoha_ppe_clear_sram_table_flow(struct airoha_ppe *ppe,uint entry_idx)
+{
+	int sram_num_entries = PPE_SRAM_NUM_ENTRIES;
+	struct airoha_foe_entry *hwe = ppe->foe;
+	
+	if(entry_idx >= sram_num_entries){
+		printk("\n Clear specific sram ppe rule - Over PPE_SRAM_NUM_ENTRIES:%d / entry_idx: %u \n",PPE_SRAM_NUM_ENTRIES,entry_idx);
+		return;
+	}else{
+
+		printk("\n Clear specific sram ppe rule - entry_idx: %u Clear hwe rule\n",entry_idx);
+		memset(&hwe[entry_idx], 0, sizeof(*hwe));
+		printk("\n Clear specific sram ppe rule - entry_idx: %u Clear SRAM rule\n",entry_idx);
+		__airoha_ppe_foe_commit_entry(ppe,entry_idx);
+		return;
+	}
+	return;
+
+}
+
+void airoha_ppe_clear_sram_table_all(struct airoha_ppe *ppe)
+{
+	int i, sram_num_entries = PPE_SRAM_NUM_ENTRIES;
+	struct airoha_foe_entry *hwe = ppe->foe;
+
+	for (i = 0; i < sram_num_entries; i++){
+		memset(&hwe[i], 0, sizeof(*hwe));
+		__airoha_ppe_foe_commit_entry(ppe, i);
+	}
+	return;
+
+}
+
+void airoha_ppe_clean_sram_table(void)
+{
+	if(glb_eth == NULL)
+		return;
+	
+	if(airoha_is_pon_sfu_mode()){
+		spin_lock_bh(&ppe_lock);
+		airoha_ppe_clear_sram_table_all(glb_eth->ppe);
+		spin_unlock_bh(&ppe_lock);
+	}
+	
+	return;
+}
+
+void airoha_ppe_delete_entry(struct airoha_ppe *ppe, struct airoha_foe_entry *hwe, uint entry_idx)
+{
+	hwe->ib1 &= ~AIROHA_FOE_IB1_BIND_STATE;
+	hwe->ib1 |= FIELD_PREP(AIROHA_FOE_IB1_BIND_STATE, AIROHA_FOE_STATE_INVALID);
+	
+	airoha_ppe_foe_commit_entry(ppe, hwe, entry_idx, false);
+	
+	return;
+
 }
 
 struct airoha_ppe_dev *airoha_ppe_get_dev(struct device *dev)
@@ -1521,6 +1735,9 @@ int airoha_ppe_init(struct airoha_eth *e
 	if (err)
 		return err;
 
+	/*clear all sram ppe table*/
+	airoha_ppe_clear_sram_table_all(ppe);
+
 	err = rhashtable_init(&eth->flow_table, &airoha_flow_table_params);
 	if (err)
 		return err;
@@ -1528,7 +1745,10 @@ int airoha_ppe_init(struct airoha_eth *e
 	err = rhashtable_init(&ppe->l2_flows, &airoha_l2_flow_table_params);
 	if (err)
 		goto error_flow_table_destroy;
-
+	rcu_assign_pointer(offload_eth_fast_tx_hook, airoha_eth_fast_tx);
+	rcu_assign_pointer(ra_sw_nat_hook_sendto_ppe, SendToPpe);
+	rcu_assign_pointer(ra_sw_nat_hook_clean_table, airoha_ppe_clean_sram_table);
+	airoha_ppe_hook_init();
 	err = airoha_ppe_debugfs_init(ppe);
 	if (err)
 		goto error_l2_flow_table_destroy;
@@ -1539,10 +1759,10 @@ error_l2_flow_table_destroy:
 	rhashtable_destroy(&ppe->l2_flows);
 error_flow_table_destroy:
 	rhashtable_destroy(&eth->flow_table);
-
 	return err;
 }
 
+
 void airoha_ppe_deinit(struct airoha_eth *eth)
 {
 	struct airoha_npu *npu;
@@ -1557,5 +1777,7 @@ void airoha_ppe_deinit(struct airoha_eth
 
 	rhashtable_destroy(&eth->ppe->l2_flows);
 	rhashtable_destroy(&eth->flow_table);
+	rcu_assign_pointer(ra_sw_nat_hook_clean_table, NULL);
+	airoha_ppe_hook_exit();
 	debugfs_remove(eth->ppe->debugfs_dir);
 }
