--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -32,7 +32,7 @@ struct airoha_eth *glb_eth;
 EXPORT_SYMBOL(glb_eth);
 extern int (*wan_speed_test_pinpong_handle_hook)(struct sk_buff*);
 extern int arht_ppe_multicast_handler(struct airoha_ppe *ppe, struct sk_buff* skb);
-extern int (*arht_xpon_igmp_sfu_enable_hook)(struct net_device *port_dev);
+//extern int (*arht_xpon_igmp_sfu_enable_hook)(struct net_device *port_dev);
 extern int arht_ppe_multicast_set_valid(struct sk_buff* skb);
 extern int airoha_is_pon_sfu_mode(void);
 int is_ethmac_nvmem_cell_available(struct airoha_eth *eth);
@@ -702,9 +702,9 @@ static void airoha_get_entry_bind(u32 ms
 				flow_e->ingress_dev_idx = ETH2;
 		}
 		airoha_ppe_check_skb(&eth->ppe->dev, q->skb, hash, false);
-		if(arht_xpon_igmp_sfu_enable_hook == NULL){
-			arht_ppe_multicast_handler(eth->ppe, q->skb);
-		}
+		//if(arht_xpon_igmp_sfu_enable_hook == NULL){
+		//	arht_ppe_multicast_handler(eth->ppe, q->skb);
+		//}
 	}
 	else{
 		if(airoha_is_pon_sfu_mode()){
