--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -1568,7 +1568,7 @@ static void airoha_qdma_init_qos(struct
 			FIELD_PREP(SLA_FAST_TICK_MASK, 25));
 	airoha_qdma_rmw(qdma, REG_SLA_TRTCM_CFG, SLA_SLOW_TICK_RATIO_MASK,
 			FIELD_PREP(SLA_SLOW_TICK_RATIO_MASK, 40));
-	airoha_dp_api_qdma_rxring_default_config(qdma);		
+	airoha_dp_api_qdma_meter_default_config(qdma);		
 }
 
 static void airoha_qdma_init_qos_stats(struct airoha_qdma *qdma)
@@ -3751,6 +3751,8 @@ static int airoha_probe(struct platform_
 	}
 	printk("\n [AIROHA_ETH]Successfully Identify CHIP ID:%d \n",eth->airoha_chip_id);
 
+	eth->meter_enable = of_property_read_bool(eth->dev->of_node, "meter-enable");
+	
 	err = dma_set_mask_and_coherent(eth->dev, DMA_BIT_MASK(32));
 	if (err) {
 		dev_err(eth->dev, "failed configuring DMA mask\n");
--- a/drivers/net/ethernet/airoha/airoha_eth.h
+++ b/drivers/net/ethernet/airoha/airoha_eth.h
@@ -34,6 +34,7 @@
 #define AIROHA_NUM_QOS_QUEUES		8
 #define AIROHA_NUM_TX_RING		32
 #define AIROHA_NUM_RX_RING		32
+#define AIROHA_NUM_RX_METER_GRP0	127
 #define AIROHA_NUM_NETDEV_TX_RINGS	(AIROHA_NUM_TX_RING + \
 					 AIROHA_MAX_NUM_CHANNELS)
 #define AIROHA_FE_MC_MAX_VLAN_TABLE	64
@@ -370,6 +371,9 @@ struct airoha_foe_mac_info {
 #define AIROHA_FOE_IB2_PSE_PORT			GENMASK(8, 5)
 #define AIROHA_FOE_IB2_NBQ			GENMASK(4, 0)
 
+
+
+
 #define AIROHA_FOE_ACTDP			GENMASK(31, 24)
 #define AIROHA_FOE_SHAPER_ID			GENMASK(23, 16)
 #define AIROHA_FOE_CHANNEL			GENMASK(15, 11)
@@ -383,6 +387,8 @@ struct airoha_foe_mac_info {
 #define AIROHA_FOE_METER_GRP3			GENMASK(8, 5)
 #define AIROHA_FOE_METER_GRP2			GENMASK(4, 0)
 
+
+
 struct airoha_foe_bridge {
 	u32 dest_mac_hi;
 
@@ -781,6 +787,7 @@ struct airoha_eth {
 	const struct airoha_eth_chip *chip;
 	struct qdma_init qdma_init;
 	bool work_started;
+	bool meter_enable;
 };
 
 u32 airoha_rr(void __iomem *base, u32 offset);
@@ -937,7 +944,7 @@ extern int airoha_debug_level;
 
 extern int airoha_upstream_fast_path_enable;
 extern int airoha_downstream_fast_path_enable;
-extern void airoha_dp_api_qdma_rxring_default_config(struct airoha_qdma *qdma);
+extern void airoha_dp_api_qdma_meter_default_config(struct airoha_qdma *qdma);
 extern int airoha_dp_api_debugfs_init(struct airoha_eth *eth);
 extern void UpdateShrinkTable(int index,const u8 *addr);
 void airoha_qdma_lan_tx(struct sk_buff *skb,u32 tag,u8 fport,int channel,int qid);
