--- a/drivers/net/ethernet/airoha/arht_dp_api.c
+++ b/drivers/net/ethernet/airoha/arht_dp_api.c
@@ -11,6 +11,7 @@
 #include <linux/etherdevice.h>
 #include <linux/if_bridge.h>
 #include <linux/rhashtable.h>
+#include <linux/version.h>
 #include <net/netfilter/nf_conntrack.h>
 #include <net/netfilter/nf_conntrack_core.h>
 #include <linux/rculist_nulls.h>
@@ -37,13 +38,135 @@ int eth_proc_exit(void);
 int arht_multicast_hwnat_state_handler_lan_only(struct airoha_ppe *ppe, struct airoha_foe_entry *hwe, unsigned int foe_index, unsigned int  port_mask, int priority);
 int arht_multicast_hwnat_state_handler_wlan_only(struct airoha_ppe *ppe, struct airoha_foe_entry *hwe, unsigned int foe_index, unsigned int  port_mask, int priority);
 int arht_multicast_hwnat_state_handler_lan_wlan(struct airoha_ppe *ppe, struct airoha_foe_entry *hwe, unsigned int foe_index, unsigned int  port_mask, int priority);
-int arht_multicast_hwnat_state_handler_xsi_only(struct airoha_ppe *ppe, struct airoha_foe_entry *hwe, unsigned int foe_index, unsigned int  port_mask, int priority);
 int arht_multicast_hwnat_state_handler_lan_hsgmii_1toN(struct airoha_ppe *ppe, struct airoha_foe_entry *hwe, unsigned int foe_index, unsigned int  port_mask, int priority);
 int arht_multicast_hwnat_state_handler_unknown(struct airoha_ppe *ppe, struct airoha_foe_entry *hwe, unsigned int foe_index, unsigned int  port_mask, int priority);
 int find_and_update_shrink_table(int select, struct hwnat_shrink_field *shrinkFieldPtr);
 int arht_multicast_handler_for_sfu(struct sk_buff* skb);
-int airoha_hwnat_clean_lan_entry_by_fport_and_channel(struct airoha_gdm_port *port, int channel);
 extern int (*arht_xpon_igmp_sfu_enable_hook)(struct net_device *port_dev);
+int qdmaSetRxRateLimitConfig(unchar ringIdx, unchar type, ushort value);
+int qdmaUpdateAllRxRateLimitValue(ushort curUnit, ushort newUnit, unchar ticker);
+int qdma_wan_tx(struct sk_buff *skb, u32 msg0, u32 msg1, struct port_info *pinfo);
+int fe_set_tx_favor_oam_enable(unchar enable);
+int mbi_hang_unlock_by_terminate(FE_Gdma_Sel_t idx);
+void fe_set_per_oq_rsv(uint port,uint oq,uint val);
+void airoha_pon_ppe_init_upd_mem(const u8 *addr);
+void UpdateShrinkTable(int index,const u8 *addr);
+void airoha_pon_set_macaddr(const u8 *addr);
+__IMEM ecnt_ret_val ecnt_qdma_hook(struct ecnt_data *in_data);
+ecnt_ret_val ecnt_fe_api_hook(struct ecnt_data *in_data);
+u32 get_fe_ppe_data(u32 reg);
+void set_fe_ppe_data(u32 reg, u32 val);
+int generalGetBucketSizeByRate(uint rateLimitValue);
+int generalSetTrtcmTokenRate(GENERAL_TrtcmModuleType_T trtcmModuleType, uint rateLimitValue, GENERAL_TrtcmRateType_T rateType, ushort index);
+int biSearchGetBucketSizeShift(uint value, uint lo, uint hi, uint unit);
+int generalSetTrtcmBucketSize(GENERAL_TrtcmModuleType_T trtcmModuleType, uint bucketSize, GENERAL_TrtcmRateType_T rateType, ushort index);
+uint generalGetTrtcmBucketSize(GENERAL_TrtcmModuleType_T trtcmModuleType, GENERAL_TrtcmRateType_T rateType, ushort index);
+uint generalGetTrtcmTokenRate(GENERAL_TrtcmModuleType_T trtcmModuleType, GENERAL_TrtcmRateType_T rateType, ushort index);
+int generalSetTrtcmMeterMode(GENERAL_TrtcmModuleType_T trtcmModuleType, GENERAL_TrtcmMeter_T meterMode, GENERAL_TrtcmRateType_T rateType, ushort index);
+int generalSetTrtcmPktMode(GENERAL_TrtcmModuleType_T trtcmModuleType, GENERAL_TrtcmPktMode_T pktMode, GENERAL_TrtcmRateType_T rateType, ushort index);
+int generalSetTrtcmTickSel(GENERAL_TrtcmModuleType_T trtcmModuleType, GENERAL_TrtcmTickSel_T tickSel, GENERAL_TrtcmRateType_T rateType, ushort index);
+uint airoha_generalGetTrtcmTokenRate(struct airoha_qdma *qdma, GENERAL_TrtcmModuleType_T trtcmModuleType,enum trtcm_mode_type rateType, ushort index);
+int airoha_dp_api_qdma_get_rxring_value(struct airoha_qdma *qdma,uint ringIdx);
+int airoha_dp_api_qdma_set_rxring_value(struct airoha_qdma *qdma,uint ringIdx,uint value);
+void airoha_dp_api_qdma_rxring_default_config(struct airoha_qdma *qdma);
+int qdmaSetTxRateLimitConfig(unchar chnlIdx, unchar type, ushort value);
+int qdmaGetTxRateLimitConfig(unchar chnlIdx, unchar type);
+int qdmaUpdateAllTxRateLimitValue(ushort curUnit, ushort newUnit, int ticksel_idx);
+int qdmaGetLimitRateMaxChnl(int ticksel_idx, int channel);
+int qdmaGetRxRateLimitConfig(unchar ringIdx, unchar type);
+u32 get_frame_engine_data(u32 reg);
+void set_frame_engine_data(u32 reg, u32 val);
+void FoeSetEntrySrcMac(uint8_t * Dst, u32 * Src_hi, u16 * Src_lo);
+void FoeSetEntryMac(uint8_t * Dst, u32 * Src_hi, u16 * Src_lo);
+int cmpMacInfo(uint8_t* Dst, uint8_t* Src);
+int airoha_set_ppe_mac(struct airoha_foe_entry *foe_entry, struct net_device *dev, char* src_mac, char*dst_mac, u16 pppid);
+__IMEM int qdmaEnableInt(struct airoha_qdma *base, uint bit, QDMA_InterruptNo_t intIdx,QDMA_IntEnable_t enableIdx);
+static inline int qdmaChecConfigDone(struct airoha_qdma *base, uint offset, uint doneBit);
+int qdmaSetTxQosScheduler(unchar channel, unchar mode, ushort weight[8]);
+int qdmaGetTxQosScheduler(unchar channel, unchar *pMode, ushort weight[8]);
+int qdmaSetQueueClose_sw(struct airoha_qdma *base, unchar channel, unchar queue);
+int qdmaSetQueueOpen_sw(struct airoha_qdma *base, unchar channel, unchar queue);
+uint qdmaIsQueueClosed_sw(unchar channel, unchar queue);
+int qdmaSetTxQosSchedulerUpstream(struct airoha_qdma *base, unchar channel, unchar mode, ushort weight[8]);
+int qdmaGetTxQosSchedulerUpstream(struct airoha_qdma *base, unchar channel, unchar *pMode, ushort weight[8]);
+int qdma_set_tx_qos_upstream(unsigned int mainType, struct ECNT_QDMA_Data *qdma_data);
+int qdma_api_set_tx_qos_upstream(unsigned int mainType, QDMA_TxQosScheduler_T *pTxQos);
+int qdma_get_tx_qos_upstream(unsigned int mainType, struct ECNT_QDMA_Data *qdma_data);
+int qdma_api_get_tx_qos_upstream(unsigned int mainType, QDMA_TxQosScheduler_T *pTxQos);
+void qdmaSetTxPeekRateMargin(struct airoha_qdma *base, QDMA_PeekRateMargin_t val);
+int qdma_clear_and_set_dbg_cntr_info(QDMA_DBG_CNTR_T *dbgCntrPtr);
+int generalGetTrtcmMode(GENERAL_TrtcmModuleType_T trtcmModuleType);
+int qdma_general_check_index_valid(GENERAL_TrtcmModuleType_T trtcmModule, GENERAL_TrtcmMode_T trtcmMode , ushort index);
+static inline int generalChecConfigDone(struct airoha_qdma *base, uint reg, uint doneBit);
+int generalGetTrtcmParaConfig(GENERAL_TrtcmModuleType_T trtcmModuleType, GENERAL_TrtcmParaType_T paraType, GENERAL_TrtcmRateType_T rateType, ushort index, uint *valueLo,uint *valueHi);
+int generalGetTrtcmTickSel(GENERAL_TrtcmModuleType_T trtcmModuleType, GENERAL_TrtcmRateType_T rateType, ushort index);
+int generalGetTrtcmFastTick(GENERAL_TrtcmModuleType_T trtcmModuleType);
+int generalGetTrtcmSlowTickRatio(GENERAL_TrtcmModuleType_T trtcmModuleType);
+uint generalGetTrtcmSlowTick(GENERAL_TrtcmModuleType_T trtcmModuleType);
+int generalGetTrtcmPktMode(GENERAL_TrtcmModuleType_T trtcmModuleType, GENERAL_TrtcmRateType_T rateType, ushort index);
+int generalSetTrtcmParaConfig(GENERAL_TrtcmModuleType_T trtcmModuleType, GENERAL_TrtcmParaType_T paraType, GENERAL_TrtcmRateType_T rateType, ushort index, uint valueLo);
+int qdmaSetChannelCfg_sw(unchar channel, uint value);
+__IMEM int qdmaDisableInt(struct airoha_qdma *base, uint bit, QDMA_InterruptNo_t intIdx,QDMA_IntEnable_t enableIdx);
+u8 get_dscp_from_skb(struct sk_buff *skb, int type);
+int airoha_is_pon_point_to_point_mode(struct sk_buff *skb);
+int sfu_packet_hash_collision_check(struct sk_buff * skb, struct airoha_foe_entry *foe_entry, u16 vlan_num);
+int airoha_check_flood_packet(struct sk_buff *skb, u16 vid, u32 tag);
+int airoha_is_pon_sfu_mode(void);
+void airoha_ppe_pon_sfu_bind(struct airoha_ppe *ppe, struct airoha_foe_entry *hwe,
+			     struct sk_buff *skb, struct port_info *pinfo, u8 fport);
+void FoeGetEntryDstMac(u8 * Dst, u32 Dst_hi, u16 Dst_lo);
+int ppe_is_multicast_entry(struct airoha_foe_entry *hwe);
+void airoha_ppe_foe_flow_update_eth_offload(struct airoha_ppe *ppe,
+					    struct sk_buff *skb, struct port_info *pinfo, u8 fport);
+int airoha_eth_transmit_packet(struct sk_buff *skb, u32 txmsg0, u32 txmsg1, struct port_info *pinfo);
+void npu_get_entry_bind(struct sk_buff *skb, u32 hash, u32 reason);
+int airoha_ppe_is_wifi2G_dev(struct net_device *dev);
+void airoha_ppe_foe_flow_update_wifi_npu_offload(struct airoha_ppe *ppe, struct sk_buff *skb, struct port_info *pinfo);
+int airoha_get_ppe_entry_state(unsigned int foe_index);
+int arht_ppe_multicast_handler(struct airoha_ppe *ppe, struct sk_buff* skb);
+int arht_ppe_multicast_set_valid(struct sk_buff* skb);
+int arht_ppe_multicast_get_valid(unsigned int foe_index);
+unsigned int arht_muliticast_get_forward_state(unsigned int port_mask, unsigned int local);
+void clear_foe_entry(struct airoha_foe_entry *hwe);
+int setMcastDstMacv4(unsigned char *mac_hdr, struct airoha_foe_entry *foe_entry);
+int setMcastDstMacv6(unsigned char *mac_hdr, struct airoha_foe_entry *foe_entry);
+int32_t PpeFillInL2Info(struct sk_buff * skb, struct airoha_foe_entry *foe_entry);
+int32_t PpeFillInL3Info(struct sk_buff * skb, struct airoha_foe_entry *foe_entry);
+int arht_set_multicast_hwnat_info(struct sk_buff* skb, struct airoha_foe_entry *hwe);
+int arht_multicast_hwnat_state_handler(struct airoha_foe_entry *hwe, unsigned int foe_index,unsigned int port_mask,unsigned int local, int priority);
+int arht_multicast_get_channel_by_stag(unsigned int stag_dp);
+int arht_multicast_hwnat_clean(unsigned int foe_index);
+int arht_multicast_hwnat_list_update(MULTICAST_HWNATENTRY_t* entry, unsigned int update_mode,unsigned int port_mask,unsigned int local);
+int arht_multicast_hwnat_list_update_lan(MULTICAST_HWNATENTRY_t* entry, unsigned int update_mode,unsigned int port_mask,unsigned int local);
+int ppeSetUpdateMemCtrl(unsigned int select, unsigned int addr, unsigned int offset);
+void ppeSetUpdMemData(unsigned int mac_data);
+int ppeSetShrinkField(int select, int index, struct hwnat_shrink_field *shrinkFieldPtr);
+void PpeClearEntryInfo(struct airoha_foe_entry *foe_entry);
+int32_t FillSpeedtestEntryInfo(struct sk_buff * skb, struct airoha_foe_entry *foe_entry);
+int isValidPpeEntry(struct sk_buff *skb, struct airoha_foe_entry *foe_entry);
+void SetSpeedtestPortInfo(struct airoha_foe_entry * foe_entry, struct airoha_ppe *ppe,struct port_info *pinfo);
+int arht_ppe_soe_offload_get_valid(struct sk_buff* skb);
+int airoha_ppe_drop_packet_handler(struct sk_buff *skb);
+void airoha_ppe_hook_init(void);
+void airoha_ppe_hook_exit(void);
+void airoha_eth_monitor_gsw_process(struct airoha_gdm_port *port);
+void airoha_eth_monitor_serdes_process(struct airoha_gdm_port *port);
+void airoha_eth_monitor(struct work_struct *work);
+int packet_hash_collision_check(struct sk_buff * skb, struct airoha_foe_entry *foe_entry, u16 vlan_num);
+void airoha_ppe_general_bind(struct airoha_ppe *ppe, struct airoha_foe_entry *hwe, struct sk_buff *skb, struct port_info *pinfo, u8 fport);
+int qdmalan_airoha_qdma_get_trtcm_param(struct airoha_qdma *qdma, int channel,
+                                       u32 addr, enum trtcm_param_type param,
+                                       enum trtcm_mode_type mode,
+                                       u32 *val_low, u32 *val_high);
+int qdma_set_qdmalan_tx_ratelimit(QDMA_TxRateLimitSet_T *txRateLimitPtr);
+int airoha_receive_hook(struct sk_buff *skb, struct airoha_qdma_desc *desc, int p, int len, u32 sptag);
+int ppe_entry_is_valid(u32 foe_entry_idx, int ring_idx);
+int airoha_get_free_lro_ring(u32 foe_entry_idx);
+int arht_general_offload_bind(struct sk_buff * skb, unsigned short type);
+struct dst_entry *arht_gen_dst_clone(struct dst_entry *dst);
+int airoha_dp_api_qdma_get_meter_value(struct airoha_qdma *qdma,uint meterIdx);
+int airoha_dp_api_qdma_set_meter_value(struct airoha_qdma *qdma,uint meterIdx,uint value);
+void airoha_dp_api_qdma_meter_default_config(struct airoha_qdma *qdma);
 QDMA_Private_T *gpQdmaPriv = NULL ;
 
 extern struct airoha_eth *glb_eth;
@@ -2128,7 +2251,7 @@ int qdma_get_tx_qos_weight(struct ECNT_Q
 
 /******************************************************************************
 ******************************************************************************/
-static int __inline__ qdmaChecConfigDone(struct airoha_qdma *base, uint offset, uint doneBit)
+static inline int qdmaChecConfigDone(struct airoha_qdma *base, uint offset, uint doneBit)
 {
 	int RETRY = 3 ;
 	volatile uint regValue = 0 ;
@@ -3602,7 +3725,7 @@ int qdma_general_check_index_valid(GENER
 
 
 
-static int __inline__ generalChecConfigDone(struct airoha_qdma *base, uint reg, uint doneBit)
+static inline int generalChecConfigDone(struct airoha_qdma *base, uint reg, uint doneBit)
 {
 	int RETRY = 3 ;
 	volatile uint regValue = 0 ;
@@ -9883,6 +10006,7 @@ void airoha_qdma_lan_tx(struct sk_buff *
 	pinfo.txq = qid;
 	pinfo.nbq = channel;
 
+#if 0
 	if (!glb_eth || !glb_eth->chip)
         return;
 
@@ -9893,6 +10017,7 @@ void airoha_qdma_lan_tx(struct sk_buff *
             break;
         }
     }
+#endif
 	
 	airoha_ppe_foe_flow_update_eth_offload(glb_eth->ppe, skb, &pinfo, fport);
 }
@@ -10291,7 +10416,7 @@ static PPE_MULTICAST_FWD_STATE_HANDLER s
 	{PPE_MULTICAST_FORWARD_STATE_WLAN_ONLY,arht_multicast_hwnat_state_handler_wlan_only},
 	{PPE_MULTICAST_FORWARD_STATE_LAN_WLAN,arht_multicast_hwnat_state_handler_lan_wlan},
 	//{PPE_MULTICAST_FORWARD_STATE_LAN_XSI,ecnt_multicast_hwnat_state_handler_lan_xsi},
-	{PPE_MULTICAST_FORWARD_STATE_XSI_ONLY,arht_multicast_hwnat_state_handler_xsi_only},
+	//{PPE_MULTICAST_FORWARD_STATE_XSI_ONLY,arht_multicast_hwnat_state_handler_xsi_only},
 	//{PPE_MULTICAST_FORWARD_STATE_XSI_WLAN,ecnt_multicast_hwnat_state_handler_xsi_wlan},	
 	{PPE_MULTICAST_FORWARD_STATE_LAN_HSGMII_1toN,arht_multicast_hwnat_state_handler_lan_hsgmii_1toN},
 	{PPE_MULTICAST_FORWARD_STATE_UNKNOWN,arht_multicast_hwnat_state_handler_unknown},
@@ -10746,6 +10871,7 @@ int arht_multicast_hwnat_state_handler_l
 }
 
 //not support, set invalid hwnat rule, make packet send to cpu
+#if 0 
 int arht_multicast_hwnat_state_handler_xsi_only(struct airoha_ppe *ppe, struct airoha_foe_entry *hwe, unsigned int foe_index, unsigned int  port_mask, int priority)
 {
 	int channel = 1;
@@ -10839,6 +10965,7 @@ int arht_multicast_hwnat_state_handler_x
 	
 	return 0;
 }
+#endif
 
 
 int arht_multicast_hwnat_state_handler_lan_hsgmii_1toN(struct airoha_ppe *ppe, struct airoha_foe_entry *hwe, unsigned int foe_index, unsigned int  port_mask, int priority)
@@ -12248,14 +12375,10 @@ static void airoha_eth_monitor_link_down
 			if (ra_sw_nat_hook_clean_table)
 				ra_sw_nat_hook_clean_table();
 
-			if(glb_eth->chip->set_channel_retire){
-				ret = glb_eth->chip->set_channel_retire(glb_eth,port);
-			}
-			else{
-				ret = airoha_fe_gdm_rls(port);
-				airoha_eth_mbi_hang_unlock_by_arbit_reset(port);
-				airoha_xsi_mac_reset(port);
-			}
+			ret = airoha_fe_gdm_rls(port);
+			airoha_eth_mbi_hang_unlock_by_arbit_reset(port);
+			airoha_xsi_mac_reset(port);
+
 			if(ret < 0 )
 				printk("\n [ARIOHA_ETH_MONITOR] Itf - %s Link Down GDM%d Rls Fail! \n",eth_dev->name,port->id);
 			else
@@ -12358,7 +12481,7 @@ void airoha_eth_monitor(struct work_stru
 	//fe_oq_stat_monitor();
 	for(p = 0; p < ARRAY_SIZE(glb_eth->ports); p++)
 	{
-		if(!glb_eth->chip->support_eth_monitor(p))
+		//if(!glb_eth->chip->support_eth_monitor(p))
 			continue;
 		
 		struct airoha_gdm_port *port = glb_eth->ports[p];
@@ -12386,7 +12509,7 @@ void airoha_eth_monitor_workqueue_init(s
 	if (!eth)
 		return;
 
-	if(eth->chip->support_eth_monitor){
+	//if(eth->chip->support_eth_monitor){
 		if(!eth->work_started)
 		{
 			eth->work_started = true;
@@ -12394,7 +12517,7 @@ void airoha_eth_monitor_workqueue_init(s
 			INIT_DELAYED_WORK(&airoha_eth_monitor_workqueue, airoha_eth_monitor);
 			schedule_delayed_work(&airoha_eth_monitor_workqueue,msecs_to_jiffies(50000));
 		}
-	}
+	//}
 }
 
 void airoha_eth_monitor_workqueue_exit(struct airoha_eth *eth)
@@ -12402,15 +12525,16 @@ void airoha_eth_monitor_workqueue_exit(s
 	if (!eth)
 		return;
 
-	if(eth->chip->support_eth_monitor){
+	//if(eth->chip->support_eth_monitor){
 		if(eth->work_started)
 		{
 			eth->work_started = false;
 			cancel_delayed_work_sync(&airoha_eth_monitor_workqueue);	
 		}
-	}
+	//}
 }
 
+#if 0
 int airoha_hwnat_clean_lan_entry_by_fport_and_channel(struct airoha_gdm_port *port, int channel)
 {
 	
@@ -12464,6 +12588,7 @@ int airoha_hwnat_clean_lan_entry_by_fpor
 	return 0;
 	
 }
+#endif
 
 static int airoha_check_pon_rx_hook(void *pMsg,struct sk_buff *skb, uint pktLen)
 {
--- a/drivers/net/ethernet/airoha/arht_dp_api_debugfs.c
+++ b/drivers/net/ethernet/airoha/arht_dp_api_debugfs.c
@@ -25,7 +25,7 @@ extern int airoha_dp_api_qdma_get_meter_
 extern int airoha_dp_api_qdma_set_meter_value(struct airoha_qdma *qdma,uint meterIdx,uint value);
 extern int packet_is_transparent_mode;
 
-ssize_t CHECK_BUF(char *debugfs_buffer, int index, int buf_size)
+static ssize_t CHECK_BUF(char *debugfs_buffer, int index, int buf_size)
 {
 	if(index >= buf_size -1)
 	{
@@ -505,7 +505,7 @@ int airoha_dp_api_debugfs_init(struct ai
 {
 	struct dentry *root;
 
-	root = debugfs_create_dir("dp_api", eth->debugfs_dir);
+	root = debugfs_create_dir("dp_api", NULL);
 	debugfs_create_file("rxring_limit_value", 0644, root, eth,
 			    &airoha_dp_api_debugfs_rxring_limit_value_fops);
 	debugfs_create_file("gemport_ratelimit", 0644, root, eth,
