--- a/drivers/net/phy/mediatek/arht_ephy_api_debugfs.c
+++ b/drivers/net/phy/mediatek/arht_ephy_api_debugfs.c
@@ -12,6 +12,10 @@
 #include <linux/uaccess.h>
 #include <linux/delay.h>
 
+void call_function(int argc, char *argv[], void *p);
+int ephy_debugfs_init(void);
+void ephy_debugfs_exit(void);
+
 extern int doPhyChkVal(int argc, char *argv[], void *p);
 extern int doPhySwVer(int argc, char *argv[], void *p);
 extern int doPhyMiiRead(int argc, char *argv[], void *p);
@@ -141,7 +145,7 @@ static ssize_t ephy_debugfs_write(struct
 
 	buf[count] = '\0';
 
-	strlcpy(local_buf, buf, sizeof(local_buf));
+	strscpy(local_buf, buf, sizeof(local_buf));
 	cur = local_buf;
 
 	while ((token = strsep(&cur, " \n")) != NULL && argc < 8)
--- a/drivers/net/phy/mediatek/tcetherphy_7583_Framework.c
+++ b/drivers/net/phy/mediatek/tcetherphy_7583_Framework.c
@@ -41,6 +41,11 @@
 
 #endif //LINUX_OS
 
+int tcPhyPortNumGet(void);
+u8 getTcPhyFlag(void);
+u32 getTcPhyRMCAPReg(int phy_add);
+bool getTcPhyEsdDetectFlag(void);
+
 extern void mtMiiRegWrite(int port_num, int reg_num, int reg_data);
 extern u32 mtMiiRegRead(u8 port_num,u8 reg_num);
 extern void mtEMiiRegWrite(u32 port_num, u32 dev_num, u32 reg_num, u32 reg_data);;
--- a/drivers/net/phy/mediatek/tcetherphy_7583_Patch.c
+++ b/drivers/net/phy/mediatek/tcetherphy_7583_Patch.c
@@ -47,6 +47,27 @@ u8 	sw_patch_flag = 1;
 u8   inital_finish_flag = 0;
 u8   disable_force_master_flag = 1; //wschung add for disable force master
 
+void PreCalSet(void);
+int phyTxAmpcomp(int show_or_comp,unsigned int phy, unsigned int TXMR_or_discrete);
+int tcPhyInit_patch(void);
+int monitor_phy_init_flag(void);
+void tcPhyChkVal(unsigned int phyAddr);
+int doPhySwVer(int argc, char *argv[], void *p);
+int sw_ability_setting(int argc, char *argv[], void *p);
+int doReadAllCalData(int argc, char *argv[], void *p);
+void sw_downshift(u8 phyaddr);
+void loopback_head_downshift_100M_main(unsigned int phyaddr,unsigned int reg1);
+void downshift_force_giga(unsigned int phyaddr,unsigned int dev_1e_reg_a2);
+void netgear_downshift_100M_main(unsigned int phy,unsigned int reg1);
+void reltek_downshift_100M_main(unsigned int phy,unsigned int reg1);
+void LDPS_NLP_sd_mask_en(u32 port);
+void LDPS_NLP_sd_mask_dis(u32 port);
+void LDPS_clear_nlp_transfer_int(u32 port);
+void LDPS_Chk_status(void);
+void en75xxGePhySwPatch(void);
+int tcPhySwPatch(void);
+void EphyMonitor(void);
+
 extern void mtMiiRegWrite(int port_num, int reg_num, int reg_data);
 extern u32 mtMiiRegRead(u8 port_num,u8 reg_num);
 extern void mtEMiiRegWrite(u32 port_num, u32 dev_num, u32 reg_num, u32 reg_data);;
--- a/drivers/net/phy/mediatek/tcetherphy_7583_PhyReg.c
+++ b/drivers/net/phy/mediatek/tcetherphy_7583_PhyReg.c
@@ -77,6 +77,10 @@ u8   tcSWVer_REG =1;
 #include <linux/device.h>
 #include <linux/module.h>
 
+int tcMiiStationRead(u32 phy_addr, u32 phy_reg);
+int tcMiiStationWrite(u32 phy_addr, u32 phy_reg, u32 phy_data);
+void mdio_cl22_write(u32 port_num,u32 dev_num,u32 reg_num,u32 reg_data);
+
 int tcMiiStationRead(u32 phy_addr, u32 phy_reg) {
 
 	//printk("Executing legacy module of mdio\n");
--- a/drivers/net/phy/mediatek/tcetherphy_7583_VendorCMD.c
+++ b/drivers/net/phy/mediatek/tcetherphy_7583_VendorCMD.c
@@ -31,6 +31,16 @@
 #include "tcetherphy.h"
 #include "tcetherphy_define.h"
 
+u8 getTcPhyLookbackFlag(void);
+u8 getTcPhyForceLinkFlag(void);
+u16 getMiiPage(char *page);
+u8 phyMbpChk(struct sk_buff* oldMbp);
+u8 phy_recv_err_check(struct sk_buff** oldMbp, u32 *frameSize);
+u32 getPhyPortMask(void);
+void phyForceEEEAllOnOffRecover(int mode);
+u32 gswPbusRead(u32 pbus_addr);
+int gswPbusWrite(u32 pbus_addr, u32 pbus_data);
+
 extern void mtMiiRegWrite(int port_num, int reg_num, int reg_data);
 extern u32 mtMiiRegRead(u8 port_num,u8 reg_num);
 extern void mtEMiiRegWrite(u32 port_num, u32 dev_num, u32 reg_num, u32 reg_data);;
