--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -3324,6 +3324,7 @@ bool airoha_is_valid_gdm_port(struct air
 static int airoha_alloc_gdm_port(struct airoha_eth *eth,
 				 struct device_node *np)
 {
+	const char *name = of_get_property(np, "label", NULL);
 	const __be32 *id_ptr = of_get_property(np, "reg", NULL);
 	struct airoha_gdm_port *port;
 	struct net_device *dev;
@@ -3356,6 +3357,8 @@ static int airoha_alloc_gdm_port(struct
 		return -ENOMEM;
 	}
 
+	if (name)
+		strlcpy(dev->name, name, IFNAMSIZ);
 	dev->netdev_ops = &airoha_netdev_ops;
 	dev->ethtool_ops = &airoha_ethtool_ops;
 	dev->max_mtu = AIROHA_MAX_MTU;
